digi/xbee/xsocket.py

```diff
@@ -0,0 +1,814 @@
+        """
+        # Check address and its contents.
+        if address is None or len(address) != 2:
+            raise ValueError("Invalid address, it must be a pair (host, port).")
+
+        port = address[1]
+        if port < 1 or port > 65535:
+            raise ValueError("Port number must be between 1 and 65535.")
+        if self.__source_port:
+            raise XBeeSocketException(status=SocketStatus.ALREADY_CONNECTED)
+
+        # If the socket is not created, create it first.
+        if self.__socket_id is None:
+            self.__create_socket()
+
+        # Create, send and check the socket create packet.
+        bind_packet = SocketBindListenPacket(self.__xbee_device.get_next_frame_id(), self.__socket_id, port)
+        response_packet = self.__xbee_device.send_packet_sync_and_get_response(bind_packet, self.__timeout)
+        self.__check_response(response_packet)
+
+        # Register the data reception from.
+        self.__register_data_received_from_callback()
+
+        # Store the source port.
+        self.__source_port = port
+
+    def listen(self, backlog=1):
+        """
+        Enables a server to accept connections.
+
+        Args:
+            backlog (Integer, optional): The number of unaccepted connections that the system will allow before refusing
+                new connections. If it's lower, it will be set to 0.
+
+        Raises:
+            XBeeSocketException: if the socket is not bound.
+        """
+        if self.__source_port is None:
+            raise XBeeSocketException(message="Socket must be bound")
+
+        self.__is_listening = True
+        self.__backlog = backlog
+
+    def accept(self):
+        """
+        Accepts a connection. The socket must be bound to an address and listening for connections.
+
+        Returns:
+            Tuple: A pair ``(conn, address)`` where ``conn`` is a new socket object usable to send and receive data on
+                the connection, and ``address`` is a pair ``(host, port)`` with the address bound to the socket on the
+                other end of the connection.
+
+        Raises:
+            XBeeException: if the connection with the XBee device is not open.
+            XBeeSocketException: if the socket is not bound or not listening.
+        """
+        if self.__source_port is None:
+            raise XBeeSocketException(message="Socket must be bound")
+        if not self.__is_listening:
+            raise XBeeSocketException(message="Socket must be listening")
+
+        lock = threading.Condition()
+        received_packet = list()
+
+        # Define the IPv4 client callback.
+        def ipv4_client_callback(packet):
+            if not isinstance(packet, SocketNewIPv4ClientPacket) or packet.socket_id != self.__socket_id:
+                return
+
+            # Add the packet to the list and notify the lock.
+            received_packet.append(packet)
+            lock.acquire()
```

Comment:
These three lines, as well as the ones used  to wait, are repeated several times, wouldn't it be better to create a pair of methods called "notify_lock(lock)" and "wait_lock(lock)"?